File: python/src/autons.py
#!/usr/bin/env python3

from vex import *
from .robot_config import chassis, brain
from vex import wait, MSEC, BRAKE

# Global variables
current_auton_selection = 0        # Current auton selection
auton_test_step = 0                # Current step in auton

def sample_auton1():
    """The first autonomous routine"""
    chassis.drive_with_voltage(3, 3)
    wait(1000, MSEC)
    chassis.stop(BRAKE)

def sample_auton2():
    """The second autonomous routine"""
    chassis.set_heading(180)  # Place the robot backwards
    chassis.drive_distance(-12, 6)
    chassis.turn_to_heading(0, 6)
    chassis.drive_distance(12, 6)

def sample_skill():
    """A long autonomous routine, e.g. skill.
    This routine is broken into steps to allow for testing of individual steps.
    This allows for easier debugging of individual parts of the long autonomous routine."""
    global auton_test_step
    
    if auton_test_step == 0:
        chassis.turn_to_heading(180)
        if not continue_auton_step():
            return  # If in test mode, stop here for testing.
    
    if auton_test_step == 1:
        chassis.drive_distance(5)
        chassis.turn_to_heading(chassis.get_heading() + 90)  # Turn right
        if not continue_auton_step():
            return  # If in test mode, stop here for testing.
    
    if auton_test_step == 2:
        chassis.turn_to_heading(chassis.get_heading() - 90)  # Turn left
        chassis.drive_distance(-5)

def run_auton_item():
    """Runs the selected autonomous routine"""
    if current_auton_selection == 0:
        sample_auton1()
    elif current_auton_selection == 1:
        sample_auton2()
    elif current_auton_selection == 2:
        sample_skill()

# The names of the autonomous routines to be displayed in the menu
auton_menu_text = [
    "auton1",
    "auton2",
    "auton_skill"
]

# ----------------------------------------------------------------------------
#                 Only change code below this line when necessary
# ----------------------------------------------------------------------------

# Total number of autons, automatically calculated based on the size of the autonMenuText array
auton_num = len(auton_menu_text)
auton_test_mode = False           # Indicates if in test mode
exit_auton_menu = False           # Flag to exit the autonomous menu
enable_end_game_timer = False     # Flag to indicate if endgame timer is enabled
END_GAME_SECONDS = 85            # Endgame reminder starts at 85 seconds

def autonomous():
    """The function is called when the autonomous period starts"""
    global exit_auton_menu, enable_end_game_timer
    
    # Exits the autonomous menu
    exit_auton_menu = True
    enable_end_game_timer = True
    
    # Runs the selected autonomous routine
    run_auton_item()

def print_menu_item():
    """This function prints the selected autonomous routine to the brain and controller screens"""
    if current_auton_selection < 0:
        print_controller_screen("quick test")
        return
    
    # Clears the brain screen
    brain.screen.clear_screen()
    
    # Sets the cursor to the third row, first column
    brain.screen.set_cursor(3, 1)
    
    # Prints the selected autonomous routine name
    brain.screen.print(auton_menu_text[current_auton_selection])
    print_controller_screen(auton_menu_text[current_auton_selection])

def continue_auton_step():
    """Continue to the next autonomous step if in test mode"""
    global auton_test_step
    
    if auton_test_mode:
        auton_test_step += 1
        return False  # Stop execution for testing
    return True  # Continue execution

def exit_auton():
    """Exit the autonomous menu"""
    # This function would handle exiting the autonomous menu
    pass

def print_controller_screen(message):
    """Print a message to the controller screen"""
    # This function would print to the controller screen
    # For now, just print to console
    print(f"Controller: {message}")

def register_auton_test_buttons():
    """Register autonomous testing buttons"""
    # This function would register buttons for testing autonomous routines
    # For now, it's a placeholder
    pass

def initialize_chassis():
    """Initialize the chassis with default settings"""
    # This function would initialize the chassis with default PID and other settings
    # For now, it's a placeholder
    pass

File: python/c++/src/autons.cpp
#include "vex.h"
int currentAutonSelection = 0;        // Current auton selection
int autonTestStep = 0;                // Current step in auton


// The first autonomous routine.
void sampleAuton1() {
  chassis.driveWithVoltage(3, 3);
  wait(1000, msec);
  chassis.stop(brake);
}

// The second autonomous routine.
void sampleAuton2() {
  chassis.setHeading(180); // place the robot backwards
  chassis.driveDistance(-12, 6);
  chassis.turnToHeading(0, 6);
  chassis.driveDistance(12, 6);
}

// A long autonomous routine, e.g. skill.
// This routine is broken into steps to allow for testing of individual steps.
// This allows for easier debugging of individual parts of the long autonomous routine.
void sampleSkill() {
  if (autonTestStep == 0) 
  {
    chassis.turnToHeading(180);
    if (!continueAutonStep()) return; // If in test mode, stop here for testing.
  }
  if (autonTestStep == 1)  
  {
    chassis.driveDistance(5);
    chassis.turnToHeading(chassis.getHeading() + 90); // Turn right
    if (!continueAutonStep()) return; // If in test mode, stop here for testing.
  } 
  if(autonTestStep == 2) 
  {
    chassis.turnToHeading(chassis.getHeading() - 90); // Turn left
    chassis.driveDistance(-5);
  }
}

// Runs the selected autonomous routine.
void runAutonItem() {
  switch (currentAutonSelection) {
  case 0:
    sampleAuton1();
    break;
  case 1:
    sampleAuton2();
    break;
  case 2:
    sampleSkill();
    break;
    }
}

// The names of the autonomous routines to be displayed in the menu.
char const * autonMenuText[] = {
  "auton1",
  "auton2",
  "auton_skill"
};



// ----------------------------------------------------------------------------
//                 Only change code below this line when necessary
// ----------------------------------------------------------------------------

int autonNum;                         // Total number of autons, automatically calculated based on the size of the autonMenuText array
bool autonTestMode = false;           // Indicates if in test mode
bool exitAutonMenu = false;           // Flag to exit the autonomous menu
bool enableEndGameTimer = false;      // Flag to indicate if endgame timer is enabled 
const int END_GAME_SECONDS = 85;      // Endgame reminder starts at 85 seconds

// The function is called when the autonomous period starts.
void autonomous(void) {
  // Exits the autonomous menu.
  exitAutonMenu = true;
  enableEndGameTimer = true;
  // Runs the selected autonomous routine.
  runAutonItem();
}

// This function prints the selected autonomous routine to the brain and controller screens.
void printMenuItem() {
  if (currentAutonSelection < 0) {
    printControllerScreen("quick test");
    return;
  }    
  // Clears the brain screen.
  Brain.Screen.clearScreen();
  // Sets the cursor to the third row, first column.
  Brain.Screen.setCursor(3, 1);
  // Prints the selected autonomous routine name.
  Brain.Screen.print("%s", autonMenuText[currentAutonSelection]);
  printControllerScreen(autonMenuText[currentAutonSelection]);
}

// This function displays the autonomous menu on the brain screen.
void showAutonMenu() {
  autonNum = sizeof(autonMenuText) / sizeof(autonMenuText[0]);
  autonTestStep = 0;

  Brain.Screen.setFont(mono30);
  printMenuItem();

  // This loop runs until the autonomous menu is exited.
  while (!exitAutonMenu) {
    // If the brain screen is pressed, cycle through the autonomous routines.
    if (Brain.Screen.pressing()) {
      // Waits until the finger is lifted up from the screen.
      while (Brain.Screen.pressing()) {
        wait(20, msec);
      }
      // Cycles through the autonomous routines.
      currentAutonSelection = (currentAutonSelection + 1) % autonNum;
      printMenuItem();
      controller1.rumble(".");
    }
    // This wait prevents the loop from using too much CPU time.
    wait(50, msec);
  }
  Brain.Screen.setFont(mono20);
}

// This function is a thread that runs in the background to remind the driver of the end game.
void endgameTimer() {
  // Waits until the end game starts.
  while (Brain.Timer.time(sec) < END_GAME_SECONDS) {
    wait(500, msec);
  }
  if (enableEndGameTimer)
  {
    printControllerScreen("end game");
    controller1.rumble("-");
  }

  // Checks the motors health every 60 seconds in drive practice
  while(true)
  {
    wait(60, seconds);
    if (!autonTestMode) checkMotors(NUMBER_OF_MOTORS);
  }
}

void exitAuton()
{
  exitAutonMenu = true;
  chassis.joystickTouched = false;
    // Clears the brain timer.
  Brain.Timer.clear();
    // Starts the end game timer thread.
  thread endgameTimer_thread = thread(endgameTimer);
  if (!chassis.joystickTouched) {
    //TODO: some macto actions
  }
  chassis.stop(coast);
}

bool setupgyro() {
  wait(100, msec);
  if (!chassis.gyro.installed()) {
    printControllerScreen("inertial sensor failure");
    controller1.rumble("---");
    wait(2, seconds);
    return false;  
  }

  chassis.gyro.calibrate(3);
  // Waits until the inertial sensor is calibrated.
  while (chassis.gyro.isCalibrating()) {
    wait(100, msec);
  }
  // Rumbles the controller to indicate that the gyro is calibrated.
  controller1.rumble(".");
  return true;
}

// This function is called before the autonomous period starts.
void pre_auton() {
  // Sets up the gyro.
  bool gyroSetupSuccess = setupgyro();

  bool motorsSetupSuccess = true;

  // Checks the motors.
  motorsSetupSuccess = checkMotors(NUMBER_OF_MOTORS);
  //set the parameters for the chassis
  setChassisDefaults();
  // Shows the autonomous menu.
  if(gyroSetupSuccess && motorsSetupSuccess) showAutonMenu();
}



// ----------------------------------------------------------------------------
//                 For testing autonomous steps
// ----------------------------------------------------------------------------
bool runAutonTest()
{
  if (!autonTestMode) return false;

  // If in test mode, run the selected autonomous routine for testing and displays the run time.
  controller(primary).rumble(".");
  double t1 = Brain.Timer.time(sec);
  runAutonItem(); 
  double t2 = Brain.Timer.time(sec);
  char timeMsg[30];
  sprintf(timeMsg, "run time: %.0f", t2-t1);
  printControllerScreen(timeMsg);
  chassis.stop(coast);
  return true;
}

bool continueAutonStep()
{
  autonTestStep++;
  if (autonTestMode) return false; // If in test mode, stop here for testing.
  return true; 
}

// This function is called when the Right button is pressed.
void buttonRightAction()
{
  if ((Brain.Timer.time(sec) < 5) && !autonTestMode) {  
    // If the button is pressed within 5 seconds of starting the program, enter test mode.
    controller1.rumble("-");
    printControllerScreen("Test Mode: ON");
    wait(1, sec);
    showAutonMenu();
    autonTestMode = true;
    return;
  } 
  if (autonTestMode)
  {
    controller1.rumble(".");
    // if in test mode, scroll through the auton menu
    currentAutonSelection = (currentAutonSelection + 1) % autonNum;
    showAutonMenu();
  }
}

void buttonLeftAction()
{
  if ((Brain.Timer.time(sec) < 5)) {
    // If the button is pressed within 5 seconds of starting the program, change the drive mode.
    changeDriveMode();
    return;
  }
  if (autonTestMode) 
  {
    // if in test mode, scroll through the auton menu
    controller1.rumble(".");
    currentAutonSelection = (currentAutonSelection - 1 + autonNum) % autonNum;
    showAutonMenu();
    return;
  }
}


void buttonDownAction()
{
  if (autonTestMode) 
  {
    // If in test mode, go to the next step.
    controller1.rumble(".");
    autonTestStep++;
    char msg[30];
    sprintf(msg, "Step: %d", autonTestStep);
    printControllerScreen(msg);
  }
}

void buttonUpAction()
{
  if (autonTestMode) 
  {
    // If in test mode, go to the previous step.
    controller1.rumble(".");
    if (autonTestStep > 0) autonTestStep--;
    char msg[30];
    sprintf(msg, "Step: %d", autonTestStep);
    printControllerScreen(msg);
  }
}

void buttonAAction()
{
  if (autonTestMode) 
  {
    // If in test mode, run the selected autonomous routine for testing and displays the run time.
    controller(primary).rumble(".");
    double t1 = Brain.Timer.time(sec);
    runAutonItem(); 
    double t2 = Brain.Timer.time(sec);
    char timeMsg[30];
    sprintf(timeMsg, "run time: %.0f", t2-t1);
    printControllerScreen(timeMsg);
    chassis.stop(coast);
  }
}

// Register the controller button callbacks for autonomous testing.
void registerAutonTestButtons()
{
  controller1.ButtonRight.pressed(buttonRightAction);
  controller1.ButtonLeft.pressed(buttonLeftAction);
  controller1.ButtonDown.pressed(buttonDownAction);
  controller1.ButtonUp.pressed(buttonUpAction);
  controller1.ButtonA.pressed(buttonAAction);
}


bool checkMotors(int motorCount, int temperatureLimit) {
  int count = 0;
  int t = 0;
  for (int i = 0; i < 20; i++) {
    motor m = motor(i);
    if (m.installed()) {
      count++;
      t = m.temperature(celsius);
      if (t > temperatureLimit) {
        controller(primary).Screen.print("motor %d is %dC           ", i + 1, t);
        controller(primary).rumble("---");
        return false;
      }
    }
  }
  if (count < motorCount) {
    controller(primary).Screen.print("%d motor is disconnected      ", motorCount - count);
    controller(primary).rumble("---");
    return false;
  }
  return true;
}

void printControllerScreen(const char* message) {
  char padded[25];
  snprintf(padded, sizeof(padded), "%-24s", message);
  controller(primary).Screen.print("%s", padded);
}
